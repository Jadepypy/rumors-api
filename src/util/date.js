const dateFormat = /^[0-9]{4}-[0-9]{2}-[0-9]{2}$/;

/**
 * Given a date string, returns a Date object if it is a valid date, raises error otherwise.

 * @param {string} dateStr        A string representation of a date in YYYY-MM-DD
 * @param {string} [name='Date']  A string used in error message

 * @throws {Error}                Error thrown if date string is invalid.
 * @return {Date}
 */
const assertAndConvertDate = (dateStr, name = 'Date') => {
  if (!dateStr.match(dateFormat)) {
    throw new Error(`${name} must be in the format of YYYY-MM-DD`);
  }
  try {
    const date = new Date(dateStr);
    if (date.toString() !== 'Invalid Date') {
      return date;
    } else {
      throw new Error();
    }
  } catch {
    throw new Error(`${name} must be a valid date in the format of YYYY-MM-DD`);
  }
};

/**
 * Given two date strings and optional maxDuration, determine if it's a valid
   date range.

 * @param {string} startDateStr             A string representation of start date in YYYY-MM-DD.
 * @param {string} endDateStr               A string representation of end date in YYYY-MM-DD.
 * @param {number} [maxDuration]            Optional maximun duration in milliseconds.
   @param {bool}   [allowFutureDate=false]  Whether the date range allows end dates in the future.

 * @throws {Error}                          Error thrown if date range is invalid.
 * @return {object}                         Corresponding Date objects for start and end date.
 */
const assertDateRange = (
  startDateStr,
  endDateStr,
  maxDuration,
  allowFutureDate = false
) => {
  if (!startDateStr || !endDateStr) {
    throw new Error('must include both start date and end date');
  }
  const startDate = assertAndConvertDate(startDateStr, 'start date');
  const endDate = assertAndConvertDate(endDateStr, 'end date');
  const duration = endDate - startDate;
  if (duration < 0) {
    throw new Error('end date cannot be earlier than start date');
  }
  if (maxDuration && duration > maxDuration) {
    // TODO: write a formator that uses the most sensible time unit to describe duration.
    throw new Error(
      `start date and end date cannot be more than ${maxDuration} apart`
    );
  }
  if (!allowFutureDate && endDate - new Date() > 0) {
    throw new Error('end date must be no later than today');
  }
  return { startDate, endDate };
};

export { assertAndConvertDate, assertDateRange };
