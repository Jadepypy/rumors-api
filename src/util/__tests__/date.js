import { assertAndConvertDate, assertDateRange } from '../date';

describe('date utils', () => {
  it('assertAndConvertDate should raise errors when given invalid date', () => {
    expect(() => assertAndConvertDate('20200101', 'start date')).toThrow(
      'start date must be in the format of YYYY-MM-DD'
    );
    expect(() => assertAndConvertDate('2020-05-35', 'start date')).toThrow(
      'start date must be a valid date in the format of YYYY-MM-DD'
    );
  });

  it('assertAndConvertDate should return Date object when given valid date string', () => {
    expect(assertAndConvertDate('2020-01-01', 'start date')).toStrictEqual(
      new Date('2020-01-01')
    );
  });

  it('assertDateRange should return Date objects if both date strings are valid', () => {
    expect(assertDateRange('2020-01-01', '2020-07-01')).toStrictEqual({
      startDate: new Date('2020-01-01'),
      endDate: new Date('2020-07-01'),
    });

    expect(
      assertDateRange('2020-01-01', '3000-07-01', undefined, true)
    ).toStrictEqual({
      startDate: new Date('2020-01-01'),
      endDate: new Date('3000-07-01'),
    });
  });

  it('assertDateRange should raise error upon invalid params', () => {
    const invalidArgs = [
      ['2020-01-01'],
      ['2020-01-01', '2019-01-01'],
      ['2019-01-01', '2020-01-01', 2592000000],
      ['3000-01-01', '3000-01-01'],
    ];
    invalidArgs.forEach(args => {
      let error;
      try {
        assertDateRange(...args);
      } catch (e) {
        error = e;
      }
      expect(error).toMatchSnapshot();
    });
  });
});
