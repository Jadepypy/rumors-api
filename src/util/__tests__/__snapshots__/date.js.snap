// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`date utils assertDateRange should raise error upon invalid params 1`] = `[Error: must include both start date and end date]`;

exports[`date utils assertDateRange should raise error upon invalid params 2`] = `[Error: end date cannot be earlier than start date]`;

exports[`date utils assertDateRange should raise error upon invalid params 3`] = `[Error: start date and end date cannot be more than 2592000000 apart]`;

exports[`date utils assertDateRange should raise error upon invalid params 4`] = `[Error: end date must be no later than today]`;
